### ISSUE TITLE
[Question] How to extract/expose the complete PyTorch computation graph (forward and backward)?

### ISSUE THREAD

**Issue by anubane** [pytorch/pytorch#69610]
How to extract the complete computation graph PyTorch  generates?

Here is my understanding:
1. The forward graph can be generated by `jit.trace` or `jit.script`
2. The backward graph is created from scratch each time `loss.backward()` is invoked in the training loop.

I am attempting to lower the computation graph generated by PyTorch into GLOW manually for some custom downstream optimization. I am not able to extract the complete computation graph at the framework level (forward AND backward).

Any help or guidance in this regard is greatly appreciated.

**Comment by mrshenli** [comment]
I recall saw @Chillee  @zou3519 posted about using the following tools. 

https://github.com/pytorch/pytorch/blob/master/torch/fx/passes/graph_drawer.py

**Comment by albanD** [comment]
This LoggingTensor https://github.com/pytorch/pytorch/blob/b10381f42db8ff259a370b6444faf641e58196a3/torch/testing/_internal/logging_tensor.py#L26 that we use for testing is also doing tracing if you're happy with getting aten ops.

**Comment by Chillee** [comment]
@anubane If you add using functorch as a dependency, you can try this

```
import torch
from functorch.compile import compiled_function, draw_graph
 
def f(x):
    return x.cos().cos()
 
def fw(f, inps):
    draw_graph(f, "forward.svg")
    return f
 
def bw(f, inps):
    draw_graph(f, "backward.svg")
    return f
    
module = Module(...)
compiled_module(module, fw_compiler=fw, bw_compiler=bw)(*inps)
```
This will extract the forwards and backwards of a given module and dump it out.

This is what @mrshenli was referring to I think.

**Comment by anubane** [comment]
These methods are very helpful for visualizing the graph. However, if I want to save the graph object to a file, how to go about doing that? How can I reload that graph object (in GLOW) to reconstruct the graph in GLOW IR?

**Comment by anubane** [comment]
> @anubane Well, how do you want to dump it to a file? You can just do something like
> 
> ```
> import torch
> from functorch.compile import compiled_function, draw_graph
>  
> def f(x):
>     return x.cos().cos()
>  
> def fw(f, inps):
>     ts_f = torch.jit.script(f)
>     torch.jit.save(ts_f)
>     return f
>  
> def bw(f, inps):
>     draw_graph(f, "backward.svg")
>     torch.jit.save(ts_f)
>     return f
>     
> module = Module(...)
> compiled_module(module, fw_compiler=fw, bw_compiler=bw)(*inps)
> ```

Naive doubt:
Will that work for saving the backward graph?

Functional doubts: 
Using `jit.save` will it not make the graph specific to a particular input instance?  

Some ML models behave differently based on the input, is it possible to capture the complete functionality - such as both static and dynamic control flows, and any other input based dynamic nature?

**Comment by Chillee** [comment]
@anubane 

> Will that work for saving the backward graph?
Yes it will work for capturing the backwards graph. Try it and see if you don't believe me :P 

> Using jit.save will it not make the graph specific to a particular input instance?
And yes, this is also true (unfortunately).

**Comment by anubane** [comment]
> @anubane If you add using functorch as a dependency, you can try this
> 
> ```
> import torch
> from functorch.compile import compiled_function, draw_graph
>  
> def f(x):
>     return x.cos().cos()
>  
> def fw(f, inps):
>     draw_graph(f, "forward.svg")
>     return f
>  
> def bw(f, inps):
>     draw_graph(f, "backward.svg")
>     return f
>     
> module = Module(...)
> compiled_module(module, fw_compiler=fw, bw_compiler=bw)(*inps)
> ```
> 
> This will extract the forwards and backwards of a given module and dump it out.
> 
> This is what @mrshenli was referring to I think.

I was not able to run this example after installing functorch with:
`pip install --user "git+https://github.com/pytorch/functorch.git@release/torch_1.10_preview"`

The module  like `draw_graph` are not available. Is the package stable yet? Is there any other way?

**Comment by ezyang** [comment]
I guess you gotta run master

**Comment by ConnollyLeon** [comment]
```python
import torch
from functorch.compile import compiled_function, draw_graph
 
def f(x):
    return x.cos().cos()
 
def fw(f, inps):
    draw_graph(f, "forward.svg")
    return f
 
def bw(f, inps):
    draw_graph(f, "backward.svg")
    return f
    
module = Module(...)
compiled_module(module, fw_compiler=fw, bw_compiler=bw)(*inps)
```


I tried to use this code to generate the complete computation graph, and here is the forward computation graph of AlexNet.
It looks awesome, but it doesn't capture the attributes like shape of the weights. 

Any suggestions on how to capture these attributes while using functorch?

![forward](https://user-images.githubusercontent.com/31810583/159839211-1e18d5de-8e75-4b5a-abdf-b42da6420793.svg)

**Comment by mgolub2** [comment]
Running into an OOM while trying this on a small transformer model - on a machine with 448GB of ram! 

What is the expected memory usage of this method? Are there alternatives to try or other settings I can alter to help reduce the memory overhead?

**Comment by Jaykumaran** [comment]
torchviz is a great package to visualize comp graph,


If you are running for first time, make sure to install graphviz.

`https://github.com/szagoruyko/pytorchviz`

**Comment by 2manikan** [comment]
Hey everyone, Ik its a bit late but I think I have something that might help: I finished building a PyTorch Graph Tracer to make debugging easier! This tool visualizes the order in which tensors are created, making it simple to understand the flow and structure of your model. It’s a solid first version, and more operationscan be added!

Feel free to test it out, share feedback or suggestions for improvement, and let me know if you find any bugs! 

The code is in this link: [2manikan/Pytorch_DAG_Visualization_Tool](https://github.com/2manikan/Pytorch_DAG_Visualization_Tool/)

Note: For now, it works by installing PyTorch, cloning the repo, and keeping all the files in the same folder. The README has more details!
